Expose HTMLMediaElement.srcObject

exoego opened the pull request:
Closes #253.

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/srcObject), `srcObject` was able to be a `MediaStream`, a `MediaSource`, a `Blob`, or a `File`.
But as of November 2017, browsers support only `MediaStream` for `srcObject`, so `srcObject` should be only `MediaStream` in this type facade.

sjrd commented on src/main/scala/org/scalajs/dom/raw/Html.scala:
```diff
@@ -3791,6 +3791,11 @@ abstract class HTMLMediaElement extends HTMLElement {
+  /**
+   * Sets or returns the object which serves as the source of the media associated with the HTMLMediaElement.
+   */
+  var srcObject: org.scalajs.dom.experimental.mediastream.MediaStream
```
Hum, even if browsers only support `MediaStream` at the moment, I believe we should use `MediaStream | MediaSource | Blob`, which is the standardized IDL type. That's because a `var` can also be *read*, and when one reads it in a future version of a browser, it could return one of the two other types.

sjrd replied:
Also, the IDL specifies `srcObject` as optional, so it should be a `js.UndefOr[MediaStream | MediaSource | Blob]`, actually.

exoego replied:
Fixed in https://github.com/scala-js/scala-js-dom/pull/365/commits/b43af8dcf87a12da2bc9950192d584b874589636

exoego commented:
Travis issue may be fixed in #368

sjrd commented:
#368 was merged. Could you rebase this PR?

exoego commented:
@sjrd 
Sorry for long absense.
I rebased and fixed formatting.
All CI jobs passed except this one: https://travis-ci.org/scala-js/scala-js-dom/jobs/646237220#L245
Could you restart only this job ?

sjrd commented on src/main/scala/org/scalajs/dom/raw/Html.scala:
```diff
@@ -3188,6 +3190,12 @@ abstract class HTMLMediaElement extends HTMLElement {
+   * Sets or returns the object which serves as the source of the media associated with the HTMLMediaElement.
+   */
+  var srcObject: js.UndefOr[
+      dom.experimental.mediastream.MediaStream | dom.experimental.mediastream.MediaSource | Blob]
```
Consider
```scala
import dom.experimental.mediastream.{MediaStream, MediaSource}
```
at the top so that you can just write
```suggestion
  var srcObject: js.UndefOr[MediaStream | MediaSource | Blob]
```

exoego replied:
Fixed in https://github.com/scala-js/scala-js-dom/pull/365/commits/053820ea6cc16097476e13349b43bee6fc910f47

sjrd commented on src/main/scala/org/scalajs/dom/experimental/mediastream/MediaStream.scala:
```diff
@@ -626,3 +626,178 @@ trait MediaTrackSupportedConstraints extends js.Object {
+@js.native
+@JSGlobal
+class MediaSource extends EventTarget {
+  def sourceBueffers: SourceBufferList = js.native
```
```suggestion
  def sourceBuffers: SourceBufferList = js.native
```

exoego replied:
Fixed in https://github.com/scala-js/scala-js-dom/pull/365/commits/328db8dedd4adfa634498ac84b364c8645ba30bb

sjrd commented on src/main/scala/org/scalajs/dom/experimental/mediastream/MediaStream.scala:
```diff
@@ -626,3 +626,178 @@ trait MediaTrackSupportedConstraints extends js.Object {
+
+  def addSourceBuffer(bufferType: String): SourceBuffer = js.native
+  def removeSourceBuffer(sourceBuffer: SourceBuffer): Unit = js.native
+  def endOfStream(error: js.UndefOr[EndOfStreamError] = ???): Unit = js.native
```
```suggestion
  def endOfStream(error: EndOfStreamError = js.native): Unit = js.native
```
better matches the IDL definition.

exoego replied:
Fixed in https://github.com/scala-js/scala-js-dom/pull/365/commits/f810f35f7005100bb590814cbfd4d248d4da5329

sjrd commented on src/main/scala/org/scalajs/dom/experimental/mediastream/MediaStream.scala:
```diff
@@ -626,3 +626,178 @@ trait MediaTrackSupportedConstraints extends js.Object {
+  var onaddtrack: js.Function1[Event, Any] = js.native
+  var onremovetrack: js.Function1[Event, Any] = js.native
+}
+@js.native
```
Please add a blank line between traits/classes/objects. Same below.

exoego replied:
Fixed in https://github.com/scala-js/scala-js-dom/pull/365/commits/4dcd3d6ab9c54f1addd9a0cfdfc52de99fe5d513

armanbilge commented:
Also assuming an [implicit approval](https://github.com/scala-js/scala-js-dom/pull/365#pullrequestreview-381048020) from @sjrd ;) 

> I just have a few small suggestions, then it's good to go.

exoego commented:
I am OK having this only on v2.

japgolly commented:
> I am OK having this only on v2.

I think there's no reason not to add this to 1.2.0 as well cos you're just adding new stuff. Our goal for 1.2.0 is to get as much in as we can without breaking bin-compat.

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/SourceBuffer[JT] var onupdate: js.Function1[Event, Any]
+experimental/mediastream/SourceBuffer[JT] var onupdateend: js.Function1[Event, Any]
+experimental/mediastream/SourceBuffer[JT] var onupdatestart: js.Function1[Event, Any]
+experimental/mediastream/SourceBuffer[JT] var timestampOffset: Double
```
Missing `trackDefaults`

japgolly replied:
Actually `trackDefaults` is deprecated and non-standard

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/SourceBuffer[JT] var onupdateend: js.Function1[Event, Any]
+experimental/mediastream/SourceBuffer[JT] var onupdatestart: js.Function1[Event, Any]
+experimental/mediastream/SourceBuffer[JT] var timestampOffset: Double
+experimental/mediastream/SourceBufferList[JT] @JSBracketAccess def apply(index: Double): SourceBuffer
```
s/Double/Int/

exoego replied:
Is `Int` OK ?
https://w3c.github.io/media-source/#sourcebufferlist says `unsigned long`.

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/SourceBufferList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], options: EventListenerOptions): Unit
+experimental/mediastream/SourceBufferList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/SourceBufferList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/SourceBufferList[JT] def length: Double
```
s/Double/Int/

exoego replied:
Is `Int` OK ?
https://w3c.github.io/media-source/#sourcebufferlist says `unsigned long`.

japgolly commented on api-reports/2_12.txt:
```diff
@@ -577,6 +577,22 @@ experimental/intl/NumberFormatOptions[JT] val minimumSignificantDigits: js.Undef
+experimental/mediastream/AudioTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/AudioTrackList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/AudioTrackList[JT] def getTrackById(id: String): js.UndefOr[AudioTrack]
+experimental/mediastream/AudioTrackList[JT] def length: Double
```
s/Double/Int/

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/SourceBufferList[JT] def removeEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/SourceBufferList[JT] var onaddsourcebuffer: js.Function1[Event, Any]
+experimental/mediastream/SourceBufferList[JT] var onremovesourcebuffer: js.Function1[Event, Any]
+experimental/mediastream/TextTrack[JT] def activeCues: js.UndefOr[TextTrackCueList]
```
I don't see in the doc anything to indicate this can be `undefined`.

exoego replied:
https://html.spec.whatwg.org/multipage/media.html#texttrack
```
readonly attribute TextTrackCueList? cues;
readonly attribute TextTrackCueList? activeCues;
```

I think `TextTrackCueList?` corresponds to `js.UndefOr[...]`

japgolly replied:
Yeah I wasn't sure either but just read through the spec and `T?` is `T | null`.
See https://heycam.github.io/webidl/#idl-nullable-type

@armanbilge FYI ↑

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrack[JT] def addCue(cue: TextTrackCue): Unit
+experimental/mediastream/TextTrack[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], options: EventListenerOptions): Unit
+experimental/mediastream/TextTrack[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/TextTrack[JT] def cues: js.UndefOr[TextTrackCueList]
```
I don't see in the doc anything to indicate this can be `undefined`.

exoego replied:
Same https://github.com/scala-js/scala-js-dom/pull/365/files#r688256324

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackCue[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/TextTrackCue[JT] def removeEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], options: EventListenerOptions): Unit
+experimental/mediastream/TextTrackCue[JT] def removeEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/TextTrackCue[JT] def track: js.UndefOr[TextTrack]
```
`| Null`, not `UndefOr`

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackCue[JT] var onexit: js.Function1[Event, Any]
+experimental/mediastream/TextTrackCue[JT] var pauseOnExit: Boolean
+experimental/mediastream/TextTrackCue[JT] var startTime: Double
+experimental/mediastream/TextTrackCueList[JT] @JSBracketAccess def apply(index: Double): TextTrackCue
```
Should be Int... or are you using Double cos we don't have native longs?

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackCue[JT] var startTime: Double
+experimental/mediastream/TextTrackCueList[JT] @JSBracketAccess def apply(index: Double): TextTrackCue
+experimental/mediastream/TextTrackCueList[JT] def getCueById(id: String): js.UndefOr[TextTrackCue]
+experimental/mediastream/TextTrackCueList[JT] def length: Double
```
Should be Int... or are you using Double cos we don't have native longs?

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackCue[JT] var pauseOnExit: Boolean
+experimental/mediastream/TextTrackCue[JT] var startTime: Double
+experimental/mediastream/TextTrackCueList[JT] @JSBracketAccess def apply(index: Double): TextTrackCue
+experimental/mediastream/TextTrackCueList[JT] def getCueById(id: String): js.UndefOr[TextTrackCue]
```
Null, not UndefOr

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackCueList[JT] @JSBracketAccess def apply(index: Double): TextTrackCue
+experimental/mediastream/TextTrackCueList[JT] def getCueById(id: String): js.UndefOr[TextTrackCue]
+experimental/mediastream/TextTrackCueList[JT] def length: Double
+experimental/mediastream/TextTrackList[JT] @JSBracketAccess def apply(index: Double): VideoTrack
```
Int

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/TextTrackList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/TextTrackList[JT] def getTrackById(id: String): js.UndefOr[TextTrack]
+experimental/mediastream/TextTrackList[JT] def length: Double
```
Int

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/TextTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], options: EventListenerOptions): Unit
+experimental/mediastream/TextTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/TextTrackList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/TextTrackList[JT] def getTrackById(id: String): js.UndefOr[TextTrack]
```
Null

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/VideoTrack[JT] val label: String
+experimental/mediastream/VideoTrack[JT] val language: String
+experimental/mediastream/VideoTrack[JT] var selected: Boolean
+experimental/mediastream/VideoTrackList[JT] @JSBracketAccess def apply(index: Double): VideoTrack
```
Int

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/VideoTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], options: EventListenerOptions): Unit
+experimental/mediastream/VideoTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/VideoTrackList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/VideoTrackList[JT] def getTrackById(id: String): js.UndefOr[VideoTrack]
```
Null

japgolly commented on api-reports/2_12.txt:
```diff
@@ -713,6 +747,101 @@ experimental/mediastream/MediaTrackSupportedConstraints[JT] var sampleRate: js.U
+experimental/mediastream/VideoTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/VideoTrackList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/VideoTrackList[JT] def getTrackById(id: String): js.UndefOr[VideoTrack]
+experimental/mediastream/VideoTrackList[JT] def length: Double
```
Int

sjrd replied:
If it says `long` or `unsigned long` in the IDL, it must be `Double` in Scala.js. Otherwise it won't be able to represent values that don't fit in an Int.

I'm not saying this is the case here. It's just a general comment since I've seen many suggestions like this change, and I'm guessing that some of them might not take the above rule into account.

armanbilge commented on api-reports/2_12.txt:
```diff
@@ -577,6 +577,22 @@ experimental/intl/NumberFormatOptions[JT] val minimumSignificantDigits: js.Undef
+experimental/mediastream/AudioTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], options: EventListenerOptions): Unit
+experimental/mediastream/AudioTrackList[JT] def addEventListener[T <: Event](`type`: String, listener: js.Function1[T, _], useCapture: Boolean?): Unit
+experimental/mediastream/AudioTrackList[JT] def dispatchEvent(evt: Event): Boolean
+experimental/mediastream/AudioTrackList[JT] def getTrackById(id: String): js.UndefOr[AudioTrack]
```
Maybe #481 should target 2.x, and for now let's keep things consistent in the 1.x series (i.e., no `X | Null`)?

exoego commented on api-reports/2_12.txt:
```diff
@@ -577,6 +577,22 @@ experimental/intl/NumberFormatOptions[JT] val minimumSignificantDigits: js.Undef
+experimental/mediastream/AudioTrack[JT] val label: String
+experimental/mediastream/AudioTrack[JT] val language: String
+experimental/mediastream/AudioTrack[JT] var enabled: Boolean
+experimental/mediastream/AudioTrackList[JT] @JSBracketAccess def apply(index: Double): AudioTrack
```
According to https://html.spec.whatwg.org/multipage/media.html#videotracklist it is `unsigned long`, so I think `Double` is more appropriate.
```
readonly attribute unsigned long length;
getter VideoTrack (unsigned long index);
VideoTrack? getTrackById(DOMString id);
readonly attribute long selectedIndex;
```

japgolly commented on api-reports/2_12.txt:
```diff
@@ -577,6 +577,22 @@ experimental/intl/NumberFormatOptions[JT] val minimumSignificantDigits: js.Undef
+experimental/mediastream/AudioTrack[JT] val kind: String
+experimental/mediastream/AudioTrack[JT] val label: String
+experimental/mediastream/AudioTrack[JT] val language: String
+experimental/mediastream/AudioTrack[JT] var enabled: Boolean
```
Actually this is in the extensions draft, not the original spec

japgolly commented:
@armanbilge This is ready to go, please merge if you're happy
